client/cos.go

```diff
@@ -0,0 +1,181 @@
+package client
+
+import (
+	"context"
+	"crypto"
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/openpubkey/openpubkey/cosigner/msgs"
+	"github.com/openpubkey/openpubkey/pktoken"
+	"github.com/openpubkey/openpubkey/util"
+	"github.com/sirupsen/logrus"
+)
+
+type CosignerProvider struct {
+	Issuer       string
+	CallbackPath string
+}
+
+func (p *CosignerProvider) GetIssuer() string {
+	return p.Issuer
+}
+
+func (c *CosignerProvider) RequestToken(ctx context.Context, signer crypto.Signer, pkt *pktoken.PKToken, redirCh chan string) (*pktoken.PKToken, error) {
+	// Find an unused port
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, fmt.Errorf("failed to bind to an available port: %w", err)
+	}
+
+	port := listener.Addr().(*net.TCPAddr).Port
+	host := fmt.Sprintf("localhost:%d", port)
+	redirectURI := fmt.Sprintf("http://%s%s", host, c.CallbackPath)
+
+	mux := http.NewServeMux()
+	server := &http.Server{
+		Addr:    host,
+		Handler: mux,
+	}
+
+	logrus.Infof("listening on http://%s/", host)
+	logrus.Info("press ctrl+c to stop")
+	go func() {
+		err := server.Serve(listener)
+		if err != nil && err != http.ErrServerClosed {
+			logrus.Error(err)
+		}
+	}()
+	defer server.Shutdown(ctx)
+
+	ch := make(chan []byte)
+	errCh := make(chan error)
+
+	// This is where we get the authcode from the Cosigner
+	mux.Handle(c.CallbackPath,
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+			// Get authcode from Cosigner via Cosigner redirecting user's browser window
+			params := r.URL.Query()
+			if _, ok := params["authcode"]; !ok {
+				errCh <- fmt.Errorf("Cosigner did not return an authcode in the URI")
+				return
+			}
+			authcode := params["authcode"][0] // This is the authcode issued by the cosigner not the OP
+
+			// Sign authcode from cosigner under PK Token and send signed authcode to Cosigner
+			sig2, err := pkt.NewSignedMessage([]byte(authcode), signer)
+			if err != nil {
+				errCh <- err
+				return
+			}
+			res, err := http.Get(fmt.Sprintf("%s/sign?sig2=%s", c.Issuer, sig2))
+			if err != nil {
+				errCh <- fmt.Errorf("error requesting MFA cosigner signature: %w", err)
+				return
+			}
+
+			// Receive response from Cosigner that has cosigner signature on PK Token
+			resBody, err := io.ReadAll(res.Body)
+			if err != nil {
+				errCh <- fmt.Errorf("error reading MFA cosigner signature response: %w", err)
+				return
+			}
+			cosSig, err := util.Base64DecodeForJWT(resBody)
+			if err != nil {
+				errCh <- fmt.Errorf("error reading MFA cosigner signature response: %w", err)
+				return
+			}
+			// Success
+			w.Write([]byte("You may now close this window"))
+			ch <- cosSig
+		}),
+	)
```

Comment:
AFAICT you can add new handlers to a mux on the fly since the mux just manages a table of handlers. That said, I agree with you that it is nicer to add the handler first and then run the server to avoid getting a hard to debug 404 if the server responds to a request before the handler is added.

Reply:
Fixed